feat(lint): lint/js/preferOptionalChaining rule

antonk52 opened the pull request:
## Summary

This PR adds a new javascript linting rule that enforces the use of optional chaining (`?.`) operator. This is currently work in progress, so far I've covered the basic cases, going to expand on it to cover more complex ones soon. Would like to get any feedback prior spending more time on this.

fixes #765

## Test Plan

I've added test cases for the rule. Going to add more complex ones as them get covered.

ematipico commented:
I wonder if we can have some more test around more nested properties?

```js
foo !== undefined && foo.bar !== undefined &&  foo.bar.lorem !== undefined && !== foo.bar.lorem.ipsum

```

antonk52 commented:
Thank you for the idea. Definitely going to add more test cases and support for longer chains of checks. Right now it only supports a single binary expression.

I am not sure if it would make sense to also include here the checks for potential optional call expressions, ie

```js
// before
if (func) func()
if (func) {func()}
if (func !== undefined) func()
func && func()

// after
func?.()
```

What do you think?

ematipico commented:
That's interesting! I think we could _think_ about it but we should be careful about certain cases, I think, especially if there are assignments:

```typescript
let myVar: number | null = null;
if (func) {
	myVar = func();
}

// after rule
let myVar = null;
myVar = func?.();
```

If the result of `func?.()` is `undefined` then we wouldn't apply the rule because it changes the type of `myVar`. So maybe we should limit the scope of the rule.

antonk52 commented:
I think the optional call expression is only valid to enforce only if there are no side effects (please see the examples in my previous comment). Ie if the function call is the only thing in the if-block-statement then it is safe to enforce the elvis usage. The examples from my previous comment are the only ones I think it makes sense to mark as lint error. I can not come up with other side effect free examples. If you can think of any please let me know.

ematipico commented:
I can't think of other examples to be honest, I think you covered most of them

sebmck commented on internal/compiler/lint/rules/js/preferOptionalChaining.ts:
```diff
@@ -0,0 +1,132 @@
+	}
+	if (
+		expr.type === "JSBinaryExpression" &&
+		["!=", "!=="].includes(expr.operator)
```
There's only a couple of cases here so it's faster to just inline the checks, there's not a lot of repetition to justify it.

```suggestion
		(expr.operator === "!=" || expr.operator === "!==)
```

antonk52 commented:
@sebmck thanks for the feedback! There isn't an eslint rule yet, but there is an open [issue](https://github.com/eslint/eslint/issues/13430) for it.

I've pushed an updated that allows to detect explicit undefined and null checks. The current implementation has a few limitations though. 

1. It does not support computed properties since they can contain any js expression and I am not sure if this should be addressed.
2. My implementation assumes that the checks start from the same base ie `a && a.b && a.b.c` gets marked as an error and `x && a && a.b && a.b.c` wont. I think this is something that also can be addressed in this PR.
3. It also assumes that the last check is member expression. I think we can also detect if it is a binary expression and shrink the left hand side. i.e. `a != null && a.b != null && a.b.c === "foobar"` becomes `a?.b?.c === "foobar"`. There can be any comparison and I think that we should preserve it. I'll try to add this too.

I'm planning to add support for optional call expressions once I'll have some time over the weekend.

There is a thing I'd like to get some feedback on, it is the member expression comparison. I've added an [util](https://github.com/romefrontend/rome/pull/1143/commits/fbb4ba307469278f85d7f06465b4101dc506c1d1#diff-5b0626d18dc5171be4eac26cb951ae20R46) to convert a fully static member expressions to an array in order to find out if two member expressions have [the same base](https://github.com/romefrontend/rome/pull/1143/commits/fbb4ba307469278f85d7f06465b4101dc506c1d1#diff-5b0626d18dc5171be4eac26cb951ae20R81). Perhaps there is better way or a more convenient one, cheers.

antonk52 commented:
Hi, I've updated the rule's implementation. It supports optional call expressions now and I've added more test cases.

I think I've discovered a potential bug that blocks this PR. For this reason I have not applied the autofixes to rome's codebase. 

It is about how rome handles the formatting of long member expressions. For instance let's look at an example

```ts
// pre autofix
a && a.b.c

// post autofix
(a?.b).c

// expected
a?.b.c
```

The current formatting leads to a runtime error. To be more specific

```ts
let a;
a?.b.c // ✅ undefined
(a?.b).c // ❌ TypeError: undefined has no properties
```

Another concern I am not sure if we want to fold the chained logical expressions with with unequal comparison. This can potentially break typescript types, [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAsiByB7AJtAvFA3qSAuKAzsAE4CWAdgOYC+A3AFA7QwCGIARhEqlBuQK4AbQVAA+Wcigj443CHXr0AZv3IBjYKUTkoSxIgAULYpRltOcgJRZ6UO1FJKoRk1ABkbqMcoA6Sah8mKABCNAwAcgJEAFtoADcWQX4IcOtMW3tM7z8pQPAIWigAeiLCEgpKDLtqKocnF0oAfhyAoNCIqNioBKSUtNqskxaIPMhCkqgIYmIAGi9XUgIoMEQCAlJ2QRAoAWFamuogA) and example

```ts
type MyNode = {type: string};
type MaybeNode = null | {node: MyNode};

function foo(arg: MaybeNode) {
    if (arg && arg.node.type !== 'some value') {
        arg.node.type; // string
    }
    if (arg?.node.type !== 'some value') {
        arg.node.type; // err, arg is possibly null
    }
}
```

sebmck commented:
Yeah I mentioned formatting being completely broken earlier:

> I know we have some bugs around how optional chaining is formatted that need to be fixed (there should be an issue), but it shouldn't effect this PR.

I don't actually think we have an issue open to track fixing it. Do you want to open one?

I think this PR is good to merge once it's rebased!

antonk52 commented:
Hi @sebmck. My bad. I must've misinterpreted your previous comment about formatting, taking it for unwanted formatting preference instead of invalid code. 

I've rebased my branch and squashed my commits into one. I've not applied the rule to rome's codebase since the formatting produces code that can fail at the runtime. I think this rule should be disabled until the formatting bug will get fixed or I can place the suppression comments where the code is invalid([example](https://github.com/romefrontend/rome/pull/1143/checks?check_run_id=1052735546#step:4:50)). Personally I prefer the first option.

Here's an issue for the bug #1168

antonk52 commented:
A few updates
- I think I've been able to solve that little issue with formatting for optionally chained member expressions.
- There were quite a few changes with the autofix and I had to manually touch a couple of places since it turns out typescript refines types differently when you use optional chaining, made an [issue](https://github.com/microsoft/TypeScript/issues/40397) for it.
- There were difficulties with formatting for optional call expression, so I had to pluck it out. I would be happy to get back to it once this PR is merged since keeping the branch updated can be a bit overwhelming. 

@sebmck do you think this is good to be merged?

antonk52 commented:
The branch is updated.

I've done some digging on why the optional call expression autofix was breaking the code. The error I keep running into is that optional call expressions get compiled into ternary expressions with parens surrounding it. ie

```js
// before
a?.b.c

// after
(a == null ? undefined : a.b).c
```

I've checked the typescript v4.0.0-beta currently used by rome and it produces a different result

```js
// tsc output
a === null || a === void 0 ? void 0 : a.b.c;
```

This leads me to believe that error lies somewhere in the process of rome bundling itself. If anyone has any hints on what might be causing it, please do share 🙂
